controller/api/public/proto_over_http_test.go

```diff
@@ -0,0 +1,411 @@
+			t.Fatalf("Expecting payload to contain message [%s], but it had [%s]", expectedMessage, actualMessage)
+		}
+	})
+
+	t.Run("Can write and read  marshalled protobuf messages", func(t *testing.T) {
+		seriesToReturn := make([]*pb.MetricSeries, 0)
+		for i := 0; i < 351; i++ {
+			seriesToReturn = append(seriesToReturn, &pb.MetricSeries{Name: pb.MetricName_LATENCY})
+		}
+
+		expectedMessage := &pb.MetricResponse{
+			Metrics: seriesToReturn,
+		}
+
+		expectedReadArray, err := proto.Marshal(expectedMessage)
+
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		serialized, err := serializeAsPayload(expectedReadArray)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		reader := bufio.NewReader(bytes.NewReader(serialized))
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		actualReadArray, err := deserializePayloadFromReader(reader)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		if !reflect.DeepEqual(actualReadArray, expectedReadArray) {
+			n := len(actualReadArray)
+			xor := make([]byte, n)
+			for i := 0; i < n; i++ {
+				xor[i] = actualReadArray[i] ^ expectedReadArray[i]
+			}
+			t.Fatalf("Expecting read byte array to be equal to written byte array, but they were different. xor: [%v]", xor)
+		}
+
+		actualMessage := &pb.MetricResponse{}
+		err = proto.Unmarshal(actualReadArray, actualMessage)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		if !reflect.DeepEqual(actualMessage, expectedMessage) {
+			t.Fatalf("Expecting payload to contain message [%s], but it had [%s]", expectedMessage, actualMessage)
+		}
+	})
+}
+
+func TestNewStreamingWriter(t *testing.T) {
+	t.Run("Returns a streaming writer if the ResponseWriter is compatible with streaming", func(t *testing.T) {
+		rawWriter := newStubResponseWriter()
+		flushableWriter, err := newStreamingWriter(rawWriter)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+
+		if flushableWriter != rawWriter {
+			t.Fatalf("Expected to return same instance of writer")
+		}
+
+		header := "Connection"
+		expectedValue := "keep-alive"
+		actualValue := rawWriter.Header().Get(header)
+		if actualValue != expectedValue {
+			t.Fatalf("Expected header [%s] to be set to [%s], but was [%s]", header, expectedValue, actualValue)
+		}
+
+		header = "Transfer-Encoding"
+		expectedValue = "chunked"
+		actualValue = rawWriter.Header().Get(header)
+		if actualValue != expectedValue {
+			t.Fatalf("Expected header [%s] to be set to [%s], but was [%s]", header, expectedValue, actualValue)
+		}
+	})
+
+	t.Run("Returns an error if writer doesnt support streaming", func(t *testing.T) {
+		_, err := newStreamingWriter(&nonStreamingResponseWriter{})
+		if err == nil {
+			t.Fatalf("Expecting error, got nothing")
+		}
+	})
+}
+
+func TestCheckIfResponseHasError(t *testing.T) {
+	t.Run("If response doesnt contain Conduit error, returns nil", func(t *testing.T) {
```

Comment:
TIOLI I'd reverse the three test descriptions inside this function to put the conditional at the end, since I think that makes them more readable.

> Returns nil if response doesn't contain Conduit error

> Returns error from body if response contains Conduit error

> Returns error if response contains Conduit error but body isn't error message